src/work/WorkManagerImpl.cpp

```diff
@@ -61,6 +61,13 @@ WorkManagerImpl::notify(std::string const& child)
         mApp.getMetrics().NewMeter({"work", "root", "failure"}, "unit").Mark();
         mChildren.erase(child);
     }
+    else if (i->second->getState() == Work::WORK_FAILURE_ABORTED)
```

Comment:
Before reviewing this PR, I opened https://github.com/stellar/stellar-core/issues/1755 as I thought semantics were already not super clean and error prone, now that we have abort(ing), we really need to formalize well what is going on, otherwise we're going to run into very strange bugs.

Also, the semantics implied here from `onAbort` don't really seem to follow what was described in https://github.com/stellar/stellar-core/issues/1706 (or at least it's unclear that it can work if `onAbort` only triggers some work).

I would recommend going back to basics: describe a state machine, its transitions and when certain callbacks (`onXYZ`) get called. Right now the `mAborting` flag makes it hard to tell which state transitions are valid vs invalid (and what is supposed to happen).

The two ways to abort are:
* somebody wants to abort some work, this causes transitions to `WORK_ABORTING` to `WORK_ABORTED` ; any work that was in flight is now complete (and was aborted if needed).
* abort in preparation for a retry, something like "decision to retry" -> `WORK_ABORTING_FOR_RETRY` (aborting work) -> `WORK_PENDING` (reset) -> `WORK_RUNNING` (run) ...